torch/csrc/api/include/torch/nn/modules/instancenorm.h

```diff
@@ -10,14 +10,38 @@ namespace nn {
 template <size_t D, typename Derived>
 // NOLINTNEXTLINE(bugprone-exception-escape)
 class InstanceNormImpl : public torch::nn::NormImplBase<D, Derived, InstanceNormOptions> {
+ private:
+  inline Tensor apply_instance_norm(const Tensor& input) {
+    return torch::nn::functional::detail::instance_norm(
+        input,
+        this->running_mean,
+        this->running_var,
+        this->weight,
+        this->bias,
+        this->is_training() || !this->options.track_running_stats(),
+        this->options.momentum(),
+        this->options.eps());
+  }
+
+  inline Tensor handle_no_batch_input(const Tensor& input) {
+    return this->apply_instance_norm(input.unsqueeze(0)).squeeze(0);
+  }
+
  public:
   using torch::nn::NormImplBase<D, Derived, InstanceNormOptions>::NormImplBase;
 
   Tensor forward(const Tensor& input) {
     this->_check_input_dim(input);
-    return torch::nn::functional::detail::instance_norm(
-      input, this->running_mean, this->running_var, this->weight, this->bias,
-      this->is_training() || !this->options.track_running_stats(), this->options.momentum(), this->options.eps());
+
+    // check if input does not have a batch-dim
+    // For InstanceNorm1D, 2D is unbatched and 3D is batched
+    // For InstanceNorm2D, 3D is unbatched and 4D is batched
+    // For InstanceNorm3D, 4D is unbatched and 5D is batched
+    if (input.dim() == D + 1) {
```

Comment:
I think this should actually be D + 2 due to the channels dim.

Also I think we need some C++-side tests to catch things like this, since the definition is separate across python / C++.

Reply:
This is actually correct, we are checking for the unbatched case here. I will add comment here.

Also, this is tested with C++ Parity test. (Changed it to D+2 and it failed locally).

Reply:
Awesome - thanks for clarifying! Looks good then :)